docs/concepts/what-is-a-workflow-definition.md

```diff
@@ -11,13 +11,97 @@ A Workflow Definition is the code that defines the constraints of a Workflow Exe
+To alleviate non-deterministic issues that arise from code changes we recommend using [Workflow Versioning](#what-is-workflow-versioning).
+
+For example, let's say we have a Workflow Definition that defines the following sequence:
+
+1. Start and wait on a Timer/sleep
+2. Spawn and wait on an Activity Execution
+3. Complete
+
+We start a Worker and spawn a Workflow Execution that uses that Workflow Definition.
+The Worker would emit the [StartTimer](/docs/concepts/what-is-a-command/#starttimer) Command and the Workflow Execution would become suspended.
+
+Before the Timer is up, we change the Workflow Definition to the following sequence:
+
+1. Spawn and wait on an Activity Execution
+2. Start and wait on a Timer/sleep
+3. Complete
+
+When the Timer fires, the next Workflow Task will cause the Workflow Function to re-execute.
+The first Command the Worker sees would be be ScheduleActivityTask Command, which wouldn't match up to the expected [TimerStarted](/docs/concepts/what-is-an-event#timerstarted] Event.
+
+The Workflow Execution would fail, and return the non-determinism error.
+
+The following are examples of some minor changes that would not take effect when re-executing a History which already contain the Events:
+
+- Changing the duration of a Timer.
+- Changing the arguments to:
+  - The Activity Options in a call to spawn an Activity Execution (local or nonlocal).
+  - The Child Workflow Options in a call to spawn a Child Workflow Execution.
+  - Call to Signal an External Workflow Execution.
+
+#### Intrinsic non-deterministic logic
+
+Intrinsic non-determinism is when a Workflow Function Execution might emit a different sequence of Commands on re-execution, regardless of whether all the input parameters are the same.
+
+For example, a Workflow Definition can not have inline logic that branches (emits a different Command sequence) based off a wall-clock or a random number.
+Here is some representative pseudocode:
+
+```text
+fn my_workflow() {
+  if system_clock().is_before("12pm") {
+    await workflow.sleep(duration_until("12pm"))
+  } else {
+    await my_afternoon_activity()
+  }
+}
+```
+
+Each SDK offers APIs that enable Workflow Definitions to have logic that gets and uses time, random numbers, and data from unreliable resources.
+When those APIs are used, the results are stored as part of the Event History, which means that a re-executed Workflow Function will issue the same sequence of Commands, even if there is branching involved.
+
+In other words, all operations that do not purely mutate the Workflow Execution's state should occur via a Temporal SDK API.
+
+### What is Workflow Versioning?
+
+The Workflow Versioning feature enables the creation of logical branching inside a Workflow Definition based on a developer specified version identifier.
+Use this feature when a Workflow Definition logic needs to be updated, but there are running Workflow Executions that currently depend on it.
```

Comment:
Probably worth still at least making reference to using a new task queue, since this is what most users do for anything other than bugfixes since the versioning API is so burdensome